src/platform/common/utils/events.ts

```diff
@@ -0,0 +1,1548 @@
+// -----------------------------------------------------------------------------------------------------------------------
+const _enableDisposeWithListenerWarning = false;
+// _enableDisposeWithListenerWarning = Boolean("TRUE"); // causes a linter warning so that it cannot be pushed
+
+// -----------------------------------------------------------------------------------------------------------------------
+// Uncomment the next line to print warnings whenever a snapshotted event is used repeatedly without cleanup.
+// See https://github.com/microsoft/vscode/issues/142851
+// -----------------------------------------------------------------------------------------------------------------------
+const _enableSnapshotPotentialLeakWarning = false;
+// _enableSnapshotPotentialLeakWarning = Boolean("TRUE"); // causes a linter warning so that it cannot be pushed
+
+export const None: Event<any> = () => BaseDisposable.None;
+
+function _addLeakageTraceLogic(options: EmitterOptions) {
+    if (_enableSnapshotPotentialLeakWarning) {
+        const { onDidAddListener: origListenerDidAdd } = options;
+        const stack = Stacktrace.create();
+        let count = 0;
+        options.onDidAddListener = () => {
+            if (++count === 2) {
+                console.warn(
+                    'snapshotted emitter LIKELY used public and SHOULD HAVE BEEN created with DisposableStore. snapshotted here'
+                );
+                stack.print();
+            }
+            origListenerDidAdd?.();
+        };
+    }
+}
+
+/**
+ * Given an event, returns another event which debounces calls and defers the listeners to a later task via a shared
+ * `setTimeout`. The event is converted into a signal (`Event<void>`) to avoid additional object creation as a
+ * result of merging events and to try prevent race conditions that could arise when using related deferred and
+ * non-deferred events.
+ *
+ * This is useful for deferring non-critical work (eg. general UI updates) to ensure it does not block critical work
+ * (eg. latency of keypress to text rendered).
+ *
+ * *NOTE* that this function returns an `Event` and it MUST be called with a `DisposableStore` whenever the returned
+ * event is accessible to "third parties", e.g the event is a public property. Otherwise a leaked listener on the
+ * returned event causes this utility to leak a listener on the original event.
+ *
+ * @param event The event source for the new event.
+ * @param disposable A disposable store to add the new EventEmitter to.
+ */
+export function defer(event: Event<unknown>, disposable?: DisposableStore | IDisposable[]): Event<void> {
+    return debounce<unknown, void>(event, () => void 0, 0, undefined, true, undefined, disposable);
+}
+
+/**
+ * Given an event, returns another event which only fires once.
+ *
+ * @param event The event source for the new event.
+ */
+export function once<T>(event: Event<T>): Event<T> {
```

Comment:
These are the things I want in this repo, a lot of these helper classes/functions have been re-invented in Jupyter, 
Its simpler to use the same code that is in VS Code (easier for everyone and its also better)